### PR TITLE
Chapter4 answers

### PR DESCRIPTION
**Pull request by aiwaverse**
### Solutions for Chapter 4

The final chapter! And the one with content I always struggled, but was very good to understand with your explanations!

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter4.hs [hunk 1]
```diff
@@ -267,7 +276,8 @@ instance Functor Maybe where
     fmap _ x = x
 @
 -}
-
+-- The type of x is a, and that implementation does not change the 
+-- type of a to b, so the return type is wrong
```
Comment (chshersh) [inline src/Chapter4.hs:280]: Absolutely correct 💯

File: src/Chapter4.hs [hunk 2]
```diff
@@ -472,10 +488,12 @@ Implement the Applicative instance for our 'Secret' data type from before.
 -}
 instance Applicative (Secret e) where
     pure :: a -> Secret e a
-    pure = error "pure Secret: Not implemented!"
+    pure = Reward
 
     (<*>) :: Secret e (a -> b) -> Secret e a -> Secret e b
-    (<*>) = error "(<*>) Secret: Not implemented!"
+    Trap e <*> _ = Trap e
+    Reward f <*> x = f <$> x
```
Comment (chshersh) [inline src/Chapter4.hs:495]: Nice work on reusing `Functor` instance for `Secret`! :world_map:

File: src/Chapter4.hs [hunk 3]
```diff
@@ -628,8 +665,10 @@ Can you implement a monad version of AND, polymorphic over any monad?
 
 🕯 HINT: Use "(>>=)", "pure" and anonymous function
 -}
+-- weird expected behavior, I personally assumed it would be the same as
+-- liftM2 (&&), but that was not the case
 andM :: (Monad m) => m Bool -> m Bool -> m Bool
-andM = error "andM: Not implemented!"
+andM b1 b2 = b1 >>= (\b -> if not b then pure b else (&&) b <$> b2)
```
Comment (chshersh) [inline src/Chapter4.hs:671]: The ordinary `&&` if lazy on the second argument. Meaning that if it meets `False` as the first argument, there is no need to calculate another arg. In `andM` we would like to keep the same behaviour, so the

```
Just False `andM` _
```

should be `Just False` :slightly_smiling_face:

File: src/Chapter4.hs [hunk 4]
```diff
@@ -673,7 +712,24 @@ Specifically,
  ❃ Implement the function to convert Tree to list
 -}
 
+-- this felt a little easy for a final boss xD
```
Comment (chshersh) [inline src/Chapter4.hs:715]: There's already an issue for that 😅 

* https://github.com/kowainik/learn4haskell/issues/56

**Review by chshersh — APPROVED** [review]
:tada: Fantastic work!
Thank you so much for taking this course! I hope you had fun :man_dancing: :microphone: 

We would really appreciate your [feedback on the course](https://docs.google.com/forms/d/e/1FAIpQLScBVhLxq5CgGnAfIGUE-fCoOUqeGkDY2HXzbT7KV2jjLOsmjQ/viewform)!

You can also ask any questions @vrom911 or me and we can discuss any topics you would like to dig into as well  :slightly_smiling_face: Don't be a stranger!

<hr>

If you would like to continue your journey to Haskell, check out our [Haskell Knowledge Map](https://kowainik.github.io/images/Haskell_Knowledge_Map.png) for the list of topics to explore 🗺️

**Comment by aiwaverse** [comment]
Thank you very much for all the help and the course.

**Pull request merged**